### PR TITLE
refactor for the /debug/endpointz changes in Istio 1.20

### PR DESCRIPTION
**Pull request by jmazzitelli**
fixes: #6510

### CODE REVIEW

**Comment by jmazzitelli** [comment]
This is how I got the test data file that is included in this PR:

```
# Install Istio 1.20 (only available as a dev build right now):
hack/istio/download-istio.sh -div 1.20-dev
hack/istio/install-istio-via-istioctl.sh -ih default -it default
# Install demo:
kubectl create namespace electronic-shop
kubectl label namespace electronic-shop istio-injection=enabled
kubectl apply -n electronic-shop -f <(curl -L https://raw.githubusercontent.com/kiali/demos/master/electronic-shop/electronic-shop.yaml)
# Get istiod data:
kubectl exec -it -n istio-system deploy/istiod -- curl http://localhost:15014/debug/endpointz | jq 
```

**Comment by jmazzitelli** [comment]
This already looks like it is close. I can see everything I would expect to see in the UI (a good graph, istio config objects, apps/services/workloads lists, etc).

But the integration tests are failing, so there looks to be some things that are still broken.

**Comment by jmazzitelli** [comment]
In order for the CI tests to pass, this diff must be done (in other words, the CI tests will fail - you have to have these changes so Istio 1.20 is used during the tests).

```diff
$ git diff
diff --git a/hack/istio/multicluster/install-primary-remote.sh b/hack/istio/multicluster/install-primary-remote.sh
index cb53962ac..8a316818e 100755
--- a/hack/istio/multicluster/install-primary-remote.sh
+++ b/hack/istio/multicluster/install-primary-remote.sh
@@ -57,7 +57,7 @@ switch_cluster "${CLUSTER1_CONTEXT}" "${CLUSTER1_USER}" "${CLUSTER1_PASS}"
 ${CLIENT_EXE} label namespace ${ISTIO_NAMESPACE} topology.istio.io/network=${NETWORK1_ID}
 
 ISTIO_INSTALL_SCRIPT="${SCRIPT_DIR}/../install-istio-via-istioctl.sh"
-${ISTIO_INSTALL_SCRIPT} --client-exe-path ${CLIENT_EXE} --cluster-name ${CLUSTER1_NAME} --istioctl ${ISTIOCTL} --istio-dir ${ISTIO_DIR} --mesh-id ${MESH_ID} --namespace ${ISTIO_NAMESPACE} --network ${NETWORK1_ID} --set values.pilot.env.EXTERNAL_ISTIOD=true --k8s-gateway-api-enabled true
+${ISTIO_INSTALL_SCRIPT} --client-exe-path ${CLIENT_EXE} --cluster-name ${CLUSTER1_NAME} --istioctl ${ISTIOCTL} --istio-dir ${ISTIO_DIR} --mesh-id ${MESH_ID} --namespace ${ISTIO_NAMESPACE} --network ${NETWORK1_ID} --set values.pilot.env.EXTERNAL_ISTIOD=true --k8s-gateway-api-enabled true -it default -ih default
 
 GEN_GATEWAY_SCRIPT="${ISTIO_DIR}/samples/multicluster/gen-eastwest-gateway.sh"
 ${GEN_GATEWAY_SCRIPT} --mesh ${MESH_ID} --cluster ${CLUSTER1_NAME} --network ${NETWORK1_ID} | ${ISTIOCTL} --context=${CLUSTER1_CONTEXT} install -y -f -
diff --git a/hack/setup-kind-in-ci.sh b/hack/setup-kind-in-ci.sh
index e85b1d337..aa4ba5676 100755
--- a/hack/setup-kind-in-ci.sh
+++ b/hack/setup-kind-in-ci.sh
@@ -93,14 +93,14 @@ if [ -n "${ISTIO_VERSION}" ]; then
 fi
 
 infomsg "Downloading istio"
-hack/istio/download-istio.sh ${DOWNLOAD_ISTIO_VERSION_ARG}
+hack/istio/download-istio.sh -div 1.20-dev
 
 setup_kind_singlecluster() {
   "${SCRIPT_DIR}"/start-kind.sh --name ci --image "${KIND_NODE_IMAGE}" 
 
   infomsg "Installing istio"
   # Apparently you can't set the requests to zero for the proxy so just setting them to some really low number.
-  hack/istio/install-istio-via-istioctl.sh --reduce-resources true --client-exe-path "$(which kubectl)" -cn "cluster-default" -mid "mesh-default" -net "network-default" -gae "true"
+  hack/istio/install-istio-via-istioctl.sh --reduce-resources true --client-exe-path "$(which kubectl)" -cn "cluster-default" -mid "mesh-default" -net "network-default" -gae "true" -ih default -it default
 
   infomsg "Pushing the images into the cluster..."
   make -e DORP="${DORP}" -e CLUSTER_TYPE="kind" -e KIND_NAME="ci" cluster-push-kiali
@@ -172,7 +172,7 @@ setup_kind_multicluster() {
   fi
 
   infomsg "Downloading istio"
-  hack/istio/download-istio.sh ${DOWNLOAD_ISTIO_VERSION_ARG}
+  hack/istio/download-istio.sh -div 1.20-dev
 
   infomsg "Cloning kiali helm-charts..."
   git clone --single-branch --branch "${TARGET_BRANCH}" https://github.com/kiali/helm-charts.git "${HELM_CHARTS_DIR}"
```

**Comment by jmazzitelli** [comment]
With the above diff in place, `hack/run-integration-tests.sh --test-suite backend` passes on my local machine.

**Comment by jmazzitelli** [comment]
With the above diff in place, `hack/run-integration-tests.sh --test-suite frontend` also passes on my local machine.

**Comment by jmazzitelli** [comment]
Leaving this in draft because we cannot merge this into master due to the fact this code will only work with Istio 1.20.

**Comment by jmazzitelli** [comment]
I'm considering removing all of this code, removing all Kiali access to istiod /debug/endpointz, and trying to obtain the needed data elsewhere. This saves us from maintaining code talking to a private/unsupported istiod API, and it saves us from loading and unmarshalling a huge amount of data for what amounts to 2 small pieces of information - a service port's "istio protocol" and a service port's "TLSMode".

What Kiali Needs Today from /debug/endpointz
===

These are the only 2 pieces of [/debug/endpointz data](https://github.com/kiali/kiali/blob/v1.73/models/port.go#L10-L21) - no other piece of data is even referenced in Kiali:
```go
	IstioProtocol string  `json:"istioProtocol"`
	TLSMode string `json:"tlsMode"`
```

Istio  Protocol
---

Note there are three "protocols" actually defined within this Port object:
```go
	Protocol      string  `json:"protocol"`
	AppProtocol   *string `json:"appProtocol,omitempty"`
	IstioProtocol string  `json:"istioProtocol"`
```

Protocol and AppProtocol are probably the same as those defined in [a k8s Service](https://kubernetes.io/docs/reference/networking/service-protocols/), but I do not know what "IstioProtocol" refers to. There are no code comments to explain this field; it is a mystery. :)

TLSMode
---

There are comments to explain TLSMode - which is a string that apparently has only 2 valid values:
* `disabled` = this endpoint should receive traffic as-is (mostly plaintext)
* `istio` = the endpoint is ready to receive Istio mTLS connections.

Where This Data Is Used in Kiali UI
===

These two pieces of data are only referenced/visualized in the Kiali UI in one place - [ServiceNetwork.tsx](https://github.com/kiali/kiali/blob/v1.73/frontend/src/pages/ServiceDetails/ServiceNetwork.tsx#L170-L178). Navigate to the "Services" left-hand nav menu, and pick some Service and go to its Overview tab. In the Network card you will see something like:
image
There you can see "mazz-protocol" (which was dummy data I added so I can visualize `IstioProtocol == 'mazz-protocol'`) and you can see the mTLS icon to visualize `TLSMode == 'istio'`

That's it. Those two pieces of string text in this ServiceNetwork card is the only place where /debug/endpointz data is used (so far as I can find). IMO, we should either (a) try to find these two pieces of data elsewhere in a public API or (b) just remove it here. Either way, it will free us from /debug/endpointz bondage - no longer would we need to maintain huge json (and thus we'd save all the dev work it would take to maintain it, as well as saving all the runtime resources that are required to retrieve the data from istiod, unmarshall it, then store the data in memory). Seems overkill to do all that just to obtain 2 pieces of data that only show up in this small corner of the Kiali UI.

**Comment by jmazzitelli** [comment]
I asked John Howard for his insight on this. He said that /debug/endpointz TLSMode is not what we think it is. He said, `TLSMode almost means "the pod has a sidecar. But in any event, can be (is, in Istio) derived from the pod/WorkloadEntry object labels.`

As for Kiali showing the lock on the service, he opines, `"Its a bit odd to use that data to give a lock icon next to a service."`

In his thinking, a better name for TLSMode would be `"This pod should handle TLS". Its the signal we use for auto-mtls to decide which endpoints are capable of handling TLS. Its one input to "will mTLS be used for this connection"`

In short, Kiali should not be exposing this internal data to Istio end-users - it is confusing at best, and no one will understand its true semantics. For sure it does NOT mean "connections to this service are always secured via mTLS".

So, because of this, and because protocol is (usually) easily derived from port name, I'm going to rip out of Kiali all its usage of this endpointz API. There is nothing in here we use or should use.

**Comment by jshaughn** [comment]
> So, because of this, and because protocol is (usually) easily derived from port name, I'm going to rip out of Kiali all its usage of this endpointz API. There is nothing in here we use or should use.

alright, so I agree it's misleading as we have it at the moment.  It's not not-valuable, it means that the endpoint is tls-capable, but it doesn't mean the traffic is definitely using mtls.  So I am +1 to eliminating the confusion and the use of this unsupported API.

**Comment by jmazzitelli** [comment]
I'm going to close this PR so I can keep my branch alive JUST in case we want to revive the ability to request and unmarshall this new /debug/endpointz API. But then I will create another PR that rips out all of this.

**Comment by jmazzitelli** [comment]
replacement PR: https://github.com/kiali/kiali/pull/6673
